### PR TITLE
Where clause anonymous functions with hints

### PR DESCRIPTION
**Pull request by xaphalanx**
Want to be able to see what object is being passed here so I can type hint it in my IDE with greater ease.

### CODE REVIEW

File: en/orm/query-builder.rst [hunk 1]
```diff
@@ -464,7 +464,7 @@ If we wanted to classify cities into SMALL, MEDIUM, or LARGE based on population
 size, we could do the following::
 
     $query = $cities->find()
-        ->where(function ($exp, $q) {
+        ->where(function (\Cake\Database\Expression\QueryExpression $exp, \Cake\ORM\Query $q) {
```
Comment (markstory) [inline en/orm/query-builder.rst]: Any chance we could use only the class name? The full name is a bit wordy.

Comment (dereuromark) [reply en/orm/query-builder.rst]: With a use statement above then?

Comment (xaphalanx) [reply en/orm/query-builder.rst]: I agree it is long and tiresome on my eyes. I decided not to because in other places on this page we use the FQCN. 

I could see doing that if we added an anonymous function definition someplace that has the FQCN and then use the shortened version else where?  Would that work for you?

**Comment by xaphalanx** [comment]
Added a block explaining the anonymous functions near the section where it explains how to use the where() method.

Shortened all the names.

**Comment by lorenzo** [comment]
pe

**Comment by lorenzo** [comment]
fat fingers, sorry.

People is gong to copy paste examples, and it will become how they work with the ORM. This increases the visual noise when reading the query.

Can you not change all the examples but justa few? I think the idea is to show what the types are, but not to make people think that this is the only way.

**Comment by dereuromark** [comment]
@lorenzo It is always easier to just remove them - even beginners should know that they are optional.
But we should in fact teach that they are useful, and we should way more use them all over the code.
The current code base - I can see that every day especially in tests - has a bad discover-ability still.
This is crucial for several reasons, including refactoring.
The performance issue is almost irrelevant here compared to the importance of proper typehinted methods and closures etc.
So I think this sets a good precedence towards what PHP7 goes anyway - we should at some point also think about typehints on protected methods, as here also the gain is immense.

File: en/orm/query-builder.rst [hunk 2]
```diff
@@ -745,7 +754,7 @@ Which will generate the following SQL looking like::
 It is also possible to build expressions using SQL functions::
 
     $query = $articles->find()
-        ->where(function ($exp, $q) {
+        ->where(function (QueryExpression , Query $q) {
```
Comment (dereuromark) [inline en/orm/query-builder.rst]: what happened to the $exp?

File: en/orm/query-builder.rst [hunk 3]
```diff
@@ -240,6 +240,15 @@ To set some basic conditions you can use the ``where()`` method::
     $query->where(['title' => 'First Post'])
         ->where(['published' => true]);
 
+You can also pass an anonymous function to the ``where()`` method. The passed anonymous function
+receives as it's first argument an instance of \Cake\Database\Expression\QueryExpression $exp
```
Comment (dereuromark) [inline en/orm/query-builder.rst]: you mean `its` here.

Comment (markstory) [reply en/orm/query-builder.rst]: See d1702f36c26f54cbbabd3860cb4ba4b438416513

File: en/orm/query-builder.rst (hunk 2 above)
Comment (markstory) [reply en/orm/query-builder.rst]: See 2b091415e1f73001d1a34dda971180e459cdc8d4

**Comment by lorenzo** [comment]
My suggestion was not about performance, just visual noise.
